rts/Net/GameServer.cpp

```diff
@@ -2608,17 +2637,76 @@ void CGameServer::CreateNewFrame(bool fromServerThread, bool fixedFrameTime)
 	}
 
 	if (!fixedFrameTime) {
-		spring_time currentTick = spring_gettime();
-		spring_time timeElapsed = currentTick - lastNewFrameTick;
+		spring_time currentTick = spring_gettime();  // Right now, in spring-ticks
+		spring_time timeElapsed = currentTick - lastNewFrameTick; // how long has elapsed since we previously called CreateNewFrame
+		
+		spring_time lastSimFrameStartTime = gu->lastSimFrameStartTime; // When the last sim frame was actually started by game thread
 
+		// Cap the max time we are behind to 200ms
 		if (timeElapsed > spring_msecs(200))
 			timeElapsed = spring_msecs(200);
 
+		// frameTimeLeft is mostly negative, and each time we add how much time has passed since last CreateNewFrame
 		frameTimeLeft += ((GAME_SPEED * 0.001f) * internalSpeed * timeElapsed.toMilliSecsf());
 		lastNewFrameTick = currentTick;
+		// If frameTimeLeft goes positive, we are overdue for ciel(frameTimeLeft) frames
 		numNewFrames = (frameTimeLeft > 0.0f)? int(math::ceil(frameTimeLeft)): 0;
 		frameTimeLeft -= numNewFrames;
 
+		#ifndef HEADLESS
+			// We add our special case, because we want to avoid issueing a new simframe right around lastSimFrameStartTime
+			float timeFromNextExpectedSimMs = (currentTick - lastSimFrameStartTime).toMilliSecsf();
+			timeFromNextExpectedSimMs = timeFromNextExpectedSimMs - (1000.0f / GAME_SPEED) / std::clamp(internalSpeed, 0.45f, 2.1f);
+			if ((numNewFrames == 1) && (serverSyncGameTiming > 0) && (vSync != 0)){
+
+				// we just inserted a single frame. if we happened to insert it in the expected window +-2ms, then hurry up by 1ms
+				timeFromNextExpectedSimMs = (currentTick - lastSimFrameStartTime).toMilliSecsf();
+
+				float simFramePeriodms =  (1000.0f / GAME_SPEED) / std::clamp(internalSpeed, 0.4f, 2.1f); // Usually 33.3 ms
```

Comment:
Where do the 0.4 and 2.1 come from, and why is it 0.4 here but 0.45 in the similar clamp above?